File: 剑指Offer/012_Exist/Exist.py
class Solution1:
    def exist(self, board: List[List[str]], word: str) -> bool:
        dx=[1,0,-1,0]
        dy=[0,1,0,-1]
        def dfs(row, col, w_idx)->bool:
            if not 0<=row<len(board) or not 0<=col<len(board[0]) or board[row][col] != word[w_idx]: return False
            if w_idx == len(word)-1: return True
            tmp = board[row][col]
            board[row][col] = '/'
            for i in range(4):
                if dfs(row+dx[i], col+dy[i], w_idx+1):
                    return True
            board[row][col] = tmp
            return False
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        if collections.Counter(word) - collections.Counter(itertools.chain(*board)):
            return False
        m, n, l = len(board), len(board[0]), len(word)
        d = ((0, 1), (1, 0), (0, -1), (-1, 0))
        def f(i, j, t):
            if t == l - 1:
                return board[i][j] == word[-1]
            v, board[i][j] = board[i][j], ''
            return any(
                0 <= (x := i + r) < m and \
                0 <= (y := j + c) < n and \
                board[x][y] == word[t + 1] and \
                f(x, y, t + 1)
                for r, c in d
            ) or board[i].__setitem__(j, v)
        return any(
            f(i, j, 0) 
            for i, j in itertools.product(range(m), range(n)) 
            if board[i][j] == word[0]
        )
    
    

File: 剑指Offer/012_Exist/Exist.cxx
#include<vector>
#include<string>
using namespace std;

class Solution {
public: 
    int dx[4] = {1,0,-1,0};
    int dy[4] = {0,1,0,-1};    
    bool exist(vector<vector<char>>& board, string word) {
        if(board.size() == 0 || board[0].size() == 0) return false;
        if(word.size() == 0) return true;

        int m=board.size(), n=board[0].size();
        
        for( int i=0; i<m;++i ){
            for(int j=0; j<n; ++j){
                if(dfs(i,j,0,word,board))
                    return true;
            }
        }
        return false;
    }

    bool dfs(int row, int col, int w_idx, string & word, vector<vector<char>>& board){
        if( !judgeEdge(row, col, board) || board[row][col]!=word[w_idx]) 
            return false;
        if(w_idx == word.size()-1){
            return true;
        }
        char tmp = board[row][col];
        board[row][col] = '/';
        for( int i=0; i<4; i++){
            int Nr=row+dx[i],Nc=col+dy[i];
            if(dfs(Nr, Nc, w_idx+1,word,board))
                return true;
        }
        board[row][col] = tmp;
        return false;
    }

    bool judgeEdge(int row, int col, vector<vector<char>>& board){
        return row >= 0 && row < board.size() && col >=0 && col < board[0].size();
    }
};